go/ql/src/experimental/CWE-525/WebCacheDeception.ql

```diff
@@ -0,0 +1,25 @@
+/*
+ * @name Web Cache Deception
+ * @description A caching system has been detected on the application and is vulnerable to web cache deception. By manipulating the URL it is possible to force the application to cache pages that are only accessible by an authenticated user. Once cached, these pages can be accessed by an unauthenticated user.
+ * @kind path-problem
+ * @problem.severity error
+ * @security-severity 9
+ * @precision high
+ * @id go/web-cache-deception
+ * @tags security
+ *       external/cwe/cwe-525
+ */
+
+import go
+
+from
+  DataFlow::CallNode httpHandleFuncCall, DataFlow::ReadNode rn, Http::HeaderWrite::Range hw,
+  DeclaredFunction f
+where
+  httpHandleFuncCall.getTarget().hasQualifiedName("net/http", "HandleFunc") and
+  httpHandleFuncCall.getArgument(0).getType().getUnderlyingType() instanceof StringType and
+  httpHandleFuncCall.getArgument(0).getStringValue().matches("%/") and
+  rn.reads(f) and
+  f.getParameter(0) = hw.getResponseWriter() and
+  hw.getHeaderName() = "cache-control"
+select httpHandleFuncCall.getArgument(0), hw.getResponseWriter()
```

Comment:
```suggestion
select httpHandleFuncCall.getArgument(0), hw.getResponseWriter()
```
Please follow [these instructions](https://codeql.github.com/docs/writing-codeql-queries/about-codeql-queries/#select-clause) on the format of the `select clause`. In particular, there needs to be a string explaining briefly what the problem is. If you want examples. search the repo for "@kind problem" to find other non-path queries and look at what message they have in their select clause.

Reply:
Thanks for your suggestions. I changed the select clause with

```
select "Wildcard Endpoint used with : " + httpHandleFuncCall.getArgument(0),
  hw.getHeaderName() + " Header is used"
```

Reply:
That doesn't match the format that the instructions specifies.

>       select element, string
> 
> element: a code element that is identified by the query, which defines where the alert is displayed.
> string: a message, which can also include links and placeholders, explaining why the alert was generated.
>

Reply:
okay i misunderstood the syntax in example [query](https://github.com/github/codeql/blob/8a8031df0ecf7e7f4f866760cce3b13651453cfc/cpp/ql/src/Critical/Unused.ql#L4)

```
select httpHandleFuncCall.getArgument(0),
  "Wildcard Endpoint used with " + httpHandleFuncCall.getArgument(0) + " and '" + hw.getHeaderName()
    + "' Header is used"
```